coderd/notifications/notifier.go

```diff
@@ -0,0 +1,234 @@
+		Count:           int32(n.cfg.LeaseCount),
+		MaxAttemptCount: int32(n.cfg.MaxSendAttempts),
+		NotifierID:      int32(n.id),
+		LeaseSeconds:    int32(n.cfg.LeasePeriod.Value().Seconds()),
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("acquire messages: %w", err)
+	}
+
+	return msgs, nil
+}
+
+// prepare has two roles:
+// 1. render the title & body templates
+// 2. build a dispatcher from the given message, payload, and these templates - to be used for delivering the notification
+func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotificationMessagesRow) (dispatch.DeliveryFunc, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	// NOTE: when we change the format of the MessagePayload, we have to bump its version and handle unmarshalling
+	// differently here based on that version.
+	var payload types.MessagePayload
+	err := json.Unmarshal(msg.Payload, &payload)
+	if err != nil {
+		return nil, xerrors.Errorf("unmarshal payload: %w", err)
+	}
+
+	handler, err := n.handlers.Resolve(msg.Method)
+	if err != nil {
+		return nil, xerrors.Errorf("resolve handler: %w", err)
+	}
+
+	var title, body string
+	if title, err = render.GoTemplate(msg.TitleTemplate, payload, nil); err != nil {
+		return nil, xerrors.Errorf("render title: %w", err)
+	}
+	if body, err = render.GoTemplate(msg.BodyTemplate, payload, nil); err != nil {
+		return nil, xerrors.Errorf("render body: %w", err)
+	}
+
+	return handler.Dispatcher(payload, title, body)
+}
+
+// deliver sends a given notification message via its defined method.
+// This method *only* returns an error when a context error occurs; any other error is interpreted as a failure to
+// deliver the notification and as such the message will be marked as failed (to later be optionally retried).
+func (n *notifier) deliver(ctx context.Context, msg database.AcquireNotificationMessagesRow, deliver dispatch.DeliveryFunc, success, failure chan<- dispatchResult) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, n.cfg.DispatchTimeout.Value())
+	defer cancel()
+	logger := n.log.With(slog.F("msg_id", msg.ID), slog.F("method", msg.Method))
+
+	retryable, err := deliver(ctx, msg.ID)
+	if err != nil {
+		// Don't try to accumulate message responses if the context has been canceled.
+		//
+		// This message's lease will expire in the store and will be requeued.
+		// It's possible this will lead to a message being delivered more than once, and that is why Stop() is preferable
+		// instead of canceling the context.
+		//
+		// In the case of backpressure (i.e. the success/failure channels are full because the database is slow),
+		// and this caused delivery timeout (CODER_NOTIFICATIONS_DISPATCH_TIMEOUT), we can't append any more updates to
+		// the channels otherwise this, too, will block.
+		if xerrors.Is(err, context.Canceled) || xerrors.Is(err, context.DeadlineExceeded) {
```

Comment:
I don't understand how success/failure channels being full can cause a delivery timeout, given that you only attempt to write to the channel after calling `deliver`

Reply:
The channels are shared across multiple notifiers.

Reply:
I understand how they can get full.  

What I don't understand is how that can cause a dispatch timeout, since you start the timeout above in this function and don't write to the channel until after delivery succeeds or fails.

```
	ctx, cancel := context.WithTimeout(ctx, n.cfg.DispatchTimeout.Value())
	defer cancel()
	logger := n.log.With(slog.F("msg_id", msg.ID), slog.F("method", msg.Method))

	retryable, err := deliver(ctx, msg.ID)
```

This code has no dependence on the state of the `success` or `failure` channels.

Reply:
> What I don't understand is how that can cause a dispatch timeout, since you start the timeout above in this function and don't write to the channel until after delivery succeeds or fails.

You're correct; I suspect I screwed this up in a refactor. IIRC I had the timeout defined a level up, and then the timeout could've been exceeded while blocked on sending an update to the channel. I'll clean this up 👍 

> Also, why is DeadlineExceeded (delivery timed out) passed back up the stack as an error? Isn't it the same as other retryable errors such that we want to send on failure channel?

Yes, it should be; I can't recall why I had this here and the tests pass without it; probably related to the refactor as well.

Thanks!